File: CapstoneProject/RentAPlace/rentaplace-ui/src/app/services/message.ts
import { Injectable } from '@angular/core';
import api from './api';

export type MessageVM = {
  messageId: number;
  senderId: number;
  senderName?: string;
  receiverId: number;
  receiverName?: string;
  propertyId?: number | null;
  messageText: string;
  createdAt: string;
  replyText?: string | null;
  replyAt?: string | null;
  isReadByReceiver?: boolean;
};

@Injectable({
  providedIn: 'root'
})
export class Message {
  private base = '/messages';

  // renter sends message to owner
  async send(dto: { receiverId: number; propertyId?: number | null; messageText: string }) {
    const res = await api.post(`${this.base}/send`, dto);
    return res.data;
  }

  // messages for current user
  async myMessages(): Promise<MessageVM[]> {
    const res = await api.get(`${this.base}/my`);
    return res.data ?? [];
  }

  // owner's messages for a property
  async forProperty(propertyId: number): Promise<MessageVM[]> {
    const res = await api.get(`${this.base}/property/${propertyId}`);
    return res.data ?? [];
  }

  // reply (owner)
  async reply(messageId: number, replyText: string) {
    const res = await api.put(`${this.base}/${messageId}/reply`, { replyText });
    return res.data;
  }

  // mark read
  async markRead(messageId: number) {
    const res = await api.put(`${this.base}/${messageId}/read`);
    return res.data;
  }
}

File: CapstoneProject/RentAPlace/RentAPlace.Api/Models/Message.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace RentAPlace.Api.Models
{
    public class Message
    {
        public int MessageId { get; set; }

        [Required]
        public int SenderId { get; set; }

        [ForeignKey("SenderId")]
        public User? Sender { get; set; }

        [Required]
        public int ReceiverId { get; set; }

        [ForeignKey("ReceiverId")]
        public User? Receiver { get; set; }

        public int? PropertyId { get; set; } // optional (not always tied to a property)

        [ForeignKey("PropertyId")]
        public Property? Property { get; set; }

        [Required, StringLength(2000)]
        public string MessageText { get; set; } = string.Empty;

        public DateTime CreatedAt { get; set; } = DateTime.Now;

        public string? ReplyText { get; set; }
        public DateTime? ReplyAt { get; set; }

        public bool IsReadByReceiver { get; set; } = false;
    }
}
